csrc/host_ir/executor.cpp

```diff
@@ -7,13 +7,157 @@
 // clang-format on
 
 #include <dynamic_transform.h>
+#include <fusion_profiler.h>
 #include <host_ir/executor.h>
+#include <instrumentation.h>
 #include <ir/utils.h>
+#include <multidevice/communication.h>
+#include <multidevice/lower_communication.h>
+#include <multidevice/utils.h>
+#include <options.h>
+#include <runtime/allocations.h>
+#include <runtime/executor_dispatch.h>
 #include <runtime/executor_kernel_arg.h>
 #include <runtime/fusion_kernel_runtime.h>
 
 namespace nvfuser {
 
+HostIrExecutor::HostIrExecutor(
+    int64_t fusion_id,
+    int64_t concrete_id,
+    int64_t runtime_id,
+    int64_t group_id)
+    : ExecutorAbstract(fusion_id, concrete_id, runtime_id, group_id),
+      communicator_(&Communicator::getInstance()) {}
+
+bool HostIrExecutor::supported(Fusion* fusion) {
+  FUSER_PERF_SCOPE("HostIrExecutor::supported");
+  if (fusion->isA<hir::HostIrContainer>()) {
+    return true;
+  }
+  std::vector<Expr*> exprs = fusion->exprs();
+  if (std::any_of(exprs.begin(), exprs.end(), [](Expr* e) {
+        return isResharding(e) && isLowerableToCommunication(e);
+      })) {
+    NVF_ERROR(
+        std::all_of(
+            exprs.begin(),
+            exprs.end(),
+            [](Expr* e) {
+              return isResharding(e) && isLowerableToCommunication(e);
+            }),
+        "Could not execute fusion as all expressions in a host IR container must be communication based at this point.");
+    return true;
+  }
+  return false;
+}
+
+void HostIrExecutor::compile(Fusion* fusion) {
+  FUSER_PERF_SCOPE("HostIrExecutor::compile");
+  NVF_ERROR(
+      supported(fusion),
+      "HostIrExecutor does not support the Fusion provided.");
+  if (isProfilerEnabled()) {
+    FusionProfiler::segment(group_id_).startCompile();
+  }
+
+  if (fusion->isA<hir::HostIrContainer>()) {
+    host_ir_container_ = std::make_unique<hir::HostIrContainer>();
+    IrCloner cloner = Fusion::copy(fusion, host_ir_container_.get());
```

Comment:
those two lines belong to if and else branches and could be move outside the block

Reply:
They couldn't so I duplicated them. I'll revisit to explain why.

Reply:
You're absolutely right. I remember duplicating this because it wasn't building, but taking them out works fine now, so I don't know what was going on then. Thanks.